src/data/fin/succ_pred.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2022 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import order.succ_pred.basic
+
+/-!
+# Successors and predecessors of `fin n`
+
+In this file, we show that `fin n` is both a `succ_order` and a `pred_order`. Note that they are
+also archimidean, but this is derived from the general instance for well-orderings as opposed
+to a specific `fin` instance.
+
+-/
+
+namespace fin
+
+instance : ∀ {n : ℕ}, succ_order (fin n)
+| 0 := by constructor; exact elim0
+| (n+1) :=
+{ succ := λ i, if i < fin.last n then i + 1 else i,
+  le_succ := λ a,
+  begin
+    cases n,
+    { exact (subsingleton.elim _ _).le },
+    split_ifs,
+    { rw [le_iff_coe_le_coe, coe_add_one_of_lt h],
+      exact (a : ℕ).le_succ },
+    { exact le_rfl }
+   end,
+  max_of_succ_le := λ a,
+  begin
+    split_ifs,
+    { rw [le_iff_coe_le_coe, coe_add_one_of_lt h],
+      intro h,
+      cases (lt_add_one _).ne (antisymm (a : ℕ).le_succ h) },
+    { rintro -,
+      rw [eq_last_of_not_lt h],
+      exact is_max_top }
+  end,
+  succ_le_of_lt := λ a b h', let h := h'.trans_le (le_last _) in
+                             by rwa [if_pos h, le_iff_coe_le_coe, coe_add_one_of_lt h],
+  le_of_lt_succ := λ a b h,
+  begin
+    split_ifs at h with hab,
+    swap, { exact h.le },
+    rw [lt_iff_coe_lt_coe, coe_add_one_of_lt hab] at h,
+    exact nat.le_of_lt_succ h
+  end }
+
+@[simp] lemma succ_eq {n : ℕ} : succ_order.succ = λ a, if a < fin.last n then a + 1 else a := rfl
+lemma succ_apply {n : ℕ} (a) : succ_order.succ a = if a < fin.last n then a + 1 else a := rfl
+
+instance : ∀ {n : ℕ}, pred_order (fin n)
+| 0 := by constructor; exact elim0
+| (n+1) :=
+{ pred := λ x, if x = 0 then 0 else x - 1,
```

Comment:
Okay, what about a constructor that asks for `¬ is_min a → b ≤ pred a ↔ a < b` and `is_min a → pred a = a`?

Reply:
is this what you had in mind?

```lean
import order.succ_pred.basic

open classical

example {α} [linear_order α] (f : α → α) (hn : ∀ {a}, ¬ is_max a → ∀ b, a < b ↔ f a ≤ b)
  (hm : ∀ a, is_max a → a = f a) : succ_order α :=
{ succ := f,
  le_succ        := λ a, by_cases (λ h, (hm a h).le) $ λ h, le_of_lt $ by simpa using (hn h a).not,
  max_of_succ_le := λ a, not_imp_not.mp $ λ h, by simpa using (hn h a).not,
  succ_le_of_lt  := λ a b, by_cases (λ h hab, hm a h ▸ hab.le) $ λ h, (hn h b).mp,
  le_of_lt_succ  := λ a b hab, by_cases (λ h, (hm b h).symm ▸ hab.le)
                                        (λ h, by simpa [hab] using (hn h a).not) }
```

(just checking before I dualise it!)

Reply:
Yup! Looks good, except maybe that I would turn `a = f a` around, and call `f` `succ`.

Reply:
I originally wanted both implicit, but that led to elaboration issues because the `by_cases` are so implicit. I don't mind turning both explicit.